text/0000-rust-crate-ownership.md

```diff
@@ -0,0 +1,158 @@
+
+Security issues in transitively intentional libraries should be handled as if they were intentional artifacts.
+
+
+### Experiment
+
+“Experiment” crates should mention they are experiments. Experiment crates may be intended to be used in a scoped sort of way; so if they are intended to be used they should be clear about what they are guaranteeing.
+
+An example of such messaging would be text like:
+
+> This crate is maintained by \[team\] as a part of an experiment around \[thingy\]. We encourage people to try to use this crate in their projects and provide feedback through \[method\], but do not guarantee long term maintenance.
+
+or, for experiments that are not intended to be used at all:
+
+> This crate is maintained by \[team\] and is an internal experiment. We do not guarantee stability or long term maintenance, please use at your own risk.
+
+Ideally, experimental crates that are published for feedback purposes will have a document to link to that lists out the purpose, rough duration, and processes of the experiment.
+
+### Deprecated
+“Deprecated” crates should contain the following text near the top of the readme/documentation:
+
+> This crate is deprecated and not intended to be used.
+
+### Placeholder
+
+“Placeholder” crates should contain the following text in their published readme/documentation:
+
+> This crate is a functionally empty crate that exists to reserve the crate name of \[tool\]. It should not be used. 
+
+In general it is better to have an empty placeholder crate published instead of reserving the crate via yanking, so that there is a readme that helps
+
+
+### Expatriated
+
+It's unclear if any action should be taken on these beyond removing any semblance of officialness (including rust-lang/foo team owners). We currently have only one such crate (`rand`).
+
+These should by and large not be considered to be "team managed" crates; this category is in this RFC for completeness to be able to talk about expatriation as an end state.
+
+## Transitions and new crates
+
+Teams should feel free to create new crates in any of these categories; however "Intentional Artifact" crates should ideally be accompanied with an RFC. As we move towards having team charters, this will also involve a charter change. Teams should notify core@rust-lang.org when they've created such crates so that the core team may track these crates and ensure this policy is applied.
+
+From time to time a team's plan for a crate may change: experiments may conclude, crates may need to be deprecated, or the team may decide to release something for wider usage.
+
+In general, teams should notify core@rust-lang.org when such a transition is being made.
+
+Any transition _away_ from "Intentional Artifact" requires an RFC.
+
+Any transition to "Intentional Artifact" should ideally be accompanied by an RFC.
+
+Expatriation should basically _never_ occur anymore, but it also requires an RFC and core team approval in case it is really necessary.
```

Comment:
Would it be possible to expand on the strength of wording here? If a team lacks bandwidth to maintain an intentional artifact but some other group of trustworthy people are willing to take it over, then doesn't that require expatriation? Or would y'all expect that the new group become a project team and be brought into the Rust project proper?

(Would it help to perhaps use `rand` as an example? e.g., If this RFC were active back when `rand` moved out of rust-lang, how would/should have it been handled instead?)

Reply:
> Would it be possible to expand on the strength of wording here? If a team lacks bandwidth to maintain an intentional artifact but some other group of trustworthy people are willing to take it over, then doesn't that require expatriation? Or would y'all expect that the new group become a project team and be brought into the Rust project proper?

I think that either it becomes a team under our governance or it gets forked; expatriation is an option but I think overall the change of guarantees is pretty major.

Reply:
Ah I think I see. It sounds like the fundamental qualification of expatriation is that the same crate name on crates.io is no longer owned by an official Rust group. But it's possible that an "intentional artifact" becomes unmaintained and some other non-official group picks up maintenance, but crucially, with a different crates.io name.

If I have that right, I think it might be worth spelling that out in the RFC text. And I agree, "should basically never happen" sounds about like the right way to handle that case.

Reply:
Moving to a set of trusted maintainers is still a marked shift from "the Rust Project supports this" in my mind - who defines "trusted"? What happens if they become "less trusted"? I think the "silent swap" is the iffy thing for me, and we can just avoid the issue+questions entirely.

For rand:
 - deliberateness around 'intentional artifact' might have prompted a discussion around whether there's real value in 'officialness' and its sustainability long term as a team-supported crate, so it may not have been official in the first place
 - if it had been official, I'd see it going via deprecation and a new crate (like rustc_serialize -> serde)